File: bindings/node/tests/flow_state.test.ts
// SPDX-License-Identifier: MPL-2.0

'use strict';

const test = require('node:test');
const assert = require('node:assert/strict');
const zlink = require('@zlink-systems/zlink');

function isConfigError(error: unknown): boolean {
  return error instanceof zlink.ConfigError;
}

test('ReceiveFlowState enum values match the C ABI', () => {
  assert.equal(zlink.ReceiveFlowState.RUNNING, 0);
  assert.equal(zlink.ReceiveFlowState.PAUSED, 1);
});

test('flow monitor event constants match the C ABI', () => {
  assert.equal(zlink.MonitorEventType.SendFlowPaused, 0x10000);
  assert.equal(zlink.MonitorEventType.SendFlowResumed, 0x20000);
  assert.equal(zlink.MonitorEventType.FlowStateStale, 0x40000);
});

test('flow monitor event flag constants match the C ABI', () => {
  assert.equal(zlink.MonitorEventFlag.ConnectionReadyEdge, 0x1);
  assert.equal(zlink.MonitorEventFlag.SendFlowWritable, 0x2);
  assert.equal(zlink.MonitorEventFlag.FlowStateStaleGeneration, 0x4);
  assert.equal(zlink.MonitorEventFlag.FlowStateStaleEpoch, 0x8);
});

test('SOCKET_MONITOR_EVENT_ALL covers the three new flow event bits', () => {
  assert.equal(
    zlink.SOCKET_MONITOR_EVENT_ALL & zlink.MonitorEventType.SendFlowPaused,
    zlink.MonitorEventType.SendFlowPaused
  );
  assert.equal(
    zlink.SOCKET_MONITOR_EVENT_ALL & zlink.MonitorEventType.SendFlowResumed,
    zlink.MonitorEventType.SendFlowResumed
  );
  assert.equal(
    zlink.SOCKET_MONITOR_EVENT_ALL & zlink.MonitorEventType.FlowStateStale,
    zlink.MonitorEventType.FlowStateStale
  );
});

test('monitor status exposes the five flow metrics, zero on a fresh PAIR socket', () => {
  const ctx = zlink.createContext();
  const pair = zlink.createPairSocket(ctx);
  const monitor = pair.monitorOpen();
  const status = monitor.status();

  for (const field of [
    'flowPausedConnections',
    'flowPauseAppliedTotal',
    'flowResumeAppliedTotal',
    'flowStateStaleTotal',
    'flowPauseDurationMs'
  ]) {
    assert.equal(typeof status[field], 'bigint', field);
    assert.equal(status[field], 0n, field);
  }

  monitor.close();
  pair.close();
  ctx.close();
});

test('monitor status detail FLOW_STATE bit matches the C ABI', () => {
  // core/include/zlink_enum.h: ZLINK_MONITOR_STATUS_DETAIL_FLOW_STATE = 1u << 5
  const ctx = zlink.createContext();
  const pair = zlink.createPairSocket(ctx);
  const monitor = pair.monitorOpen();
  const status = monitor.status();

  assert.equal(typeof status.isFlowStateDetailPopulated, 'function');
  assert.equal(
    status.isFlowStateDetailPopulated(),
    (status.detailFlags & (1 << 5)) !== 0
  );

  monitor.close();
  pair.close();
  ctx.close();
});

test('setReceiveFlowState succeeds on DEALER/ROUTER and repeat calls are idempotent', () => {
  const ctx = zlink.createContext();
  const router = zlink.createRouterSocket(ctx);
  const dealer = zlink.createDealerSocket(ctx);

  router.bind('inproc://flow-state-dealer-router-idempotent');
  dealer.connect('inproc://flow-state-dealer-router-idempotent');

  router.setReceiveFlowState(zlink.ReceiveFlowState.PAUSED);
  router.setReceiveFlowState(zlink.ReceiveFlowState.PAUSED);
  router.setReceiveFlowState(zlink.ReceiveFlowState.RUNNING);
  router.setReceiveFlowState(zlink.ReceiveFlowState.RUNNING);

  dealer.setReceiveFlowState(zlink.ReceiveFlowState.PAUSED);
  dealer.setReceiveFlowState(zlink.ReceiveFlowState.PAUSED);
  dealer.setReceiveFlowState(zlink.ReceiveFlowState.RUNNING);
  dealer.setReceiveFlowState(zlink.ReceiveFlowState.RUNNING);

  dealer.close();
  router.close();
  ctx.close();
});

test('setReceiveFlowState reports NotSupported on PAIR, PUB/SUB family, and STREAM', () => {
  const ctx = zlink.createContext();
  const pair = zlink.createPairSocket(ctx);
  const pub = zlink.createPubSocket(ctx);
  const sub = zlink.createSubSocket(ctx);
  const xpub = zlink.createXPubSocket(ctx);
  const xsub = zlink.createXSubSocket(ctx);
  const stream = zlink.createStreamSocket(ctx);

  for (const socket of [pair, pub, sub, xpub, xsub, stream]) {
    assert.throws(
      () => socket.setReceiveFlowState(zlink.ReceiveFlowState.PAUSED),
      (error: unknown) => isConfigError(error)
        && (error as InstanceType<typeof zlink.ConfigError>).result === zlink.ConfigResult.NotSupported
    );
  }

  // Existing send/recv behavior on an unsupported socket type is unaffected
  // by the rejected call.
  pair.bind('inproc://flow-state-pair-unaffected');
  const peer = zlink.createPairSocket(ctx);
  peer.connect('inproc://flow-state-pair-unaffected');
  peer.send().message('still-works').submit();
  const received = new zlink.Received();
  pair.recv(received);
  assert.equal(received.parts[0].data().toString(), 'still-works');

  peer.close();
  stream.close();
  xsub.close();
  xpub.close();
  sub.close();
  pub.close();
  pair.close();
  ctx.close();
});

test('setReceiveFlowState maps invalid handle and invalid argument per the config error policy', () => {
  const ctx = zlink.createContext();
  const router = zlink.createRouterSocket(ctx);
  router.close();

  assert.throws(
    () => router.setReceiveFlowState(zlink.ReceiveFlowState.RUNNING),
    (error: unknown) => isConfigError(error)
      && (error as InstanceType<typeof zlink.ConfigError>).result === zlink.ConfigResult.InvalidHandle
  );

  const dealer = zlink.createDealerSocket(ctx);
  for (const outOfRange of [2, -1, 999]) {
    assert.throws(
      () => dealer.setReceiveFlowState(outOfRange as unknown as typeof zlink.ReceiveFlowState[keyof typeof zlink.ReceiveFlowState]),
      (error: unknown) => isConfigError(error)
        && (error as InstanceType<typeof zlink.ConfigError>).result === zlink.ConfigResult.InvalidArgument
    );
  }

  dealer.close();
  ctx.close();
});

test('setReceiveFlowState after close observes only a bounded close-related outcome', () => {
  // Node's single JS-thread execution model cannot produce a genuine
  // concurrent close/config race the way thread-based bindings (C++, Go,
  // Java, Python, Rust) do. This test instead exercises the deterministic
  // sequential ordering documented in worklog/stage7-c-api.md §1.2 as one of
  // the two observable "close won" shapes: a fully torn-down handle reports
  // InvalidHandle. Repeated to confirm it never throws anything else or
  // corrupts state.
  for (let i = 0; i < 20; i += 1) {
    const ctx = zlink.createContext();
    const router = zlink.createRouterSocket(ctx);
    router.bind(`inproc://flow-state-close-race-${i}`);
    router.close();

    let observedResult: number | undefined;
    try {
      router.setReceiveFlowState(zlink.ReceiveFlowState.RUNNING);
      observedResult = zlink.ConfigResult.Ok;
    } catch (error) {
      assert.ok(isConfigError(error), 'expected a ConfigError');
      const result = (error as InstanceType<typeof zlink.ConfigError>).result;
      assert.ok(
        result === zlink.ConfigResult.InvalidHandle || result === zlink.ConfigResult.InvalidState,
        `unexpected result ${result}`
      );
      observedResult = result;
    }
    assert.notEqual(observedResult, undefined);
    ctx.close();
  }
});

test('public surface has no flow-frame receive/encode API or PAUSE-bypass send variant', () => {
  const ctx = zlink.createContext();
  const sockets = [
    zlink.createPairSocket(ctx),
    zlink.createPubSocket(ctx),
    zlink.createSubSocket(ctx),
    zlink.createXPubSocket(ctx),
    zlink.createXSubSocket(ctx),
    zlink.createDealerSocket(ctx),
    zlink.createRouterSocket(ctx),
    zlink.createStreamSocket(ctx),
  ];

  const forbiddenSubstrings = [
    'flowframe',
    'flow_frame',
    'encodeflow',
    'decodeflow',
    'receiveflowframe',
    'sendflowframe',
    'pausebypass',
    'bypasspause',
  ];

  for (const socket of sockets) {
    const methodNames = new Set<string>();
    let proto = Object.getPrototypeOf(socket);
    while (proto && proto !== Object.prototype) {
      for (const name of Object.getOwnPropertyNames(proto)) {
        methodNames.add(name);
      }
      proto = Object.getPrototypeOf(proto);
    }
    for (const name of methodNames) {
      const lower = name.toLowerCase();
      for (const forbidden of forbiddenSubstrings) {
        assert.ok(
          !lower.includes(forbidden),
          `socket exposes forbidden flow API member: ${name}`
        );
      }
    }
    // The only public member whose name contains "flow" is the state setter.
    const flowMembers = [...methodNames].filter((name) => name.toLowerCase().includes('flow'));
    assert.deepEqual(flowMembers, ['setReceiveFlowState']);
    socket.close();
  }

  for (const exportName of Object.keys(zlink)) {
    const lower = exportName.toLowerCase();
    for (const forbidden of forbiddenSubstrings) {
      assert.ok(
        !lower.includes(forbidden),
        `zlink package exposes forbidden flow API export: ${exportName}`
      );
    }
  }

  ctx.close();
});

test('existing HWM options and DEALER/ROUTER traffic are unchanged by receive-flow-state calls', async () => {
  const ctx = zlink.createContext();
  ctx.options.autoHwmEnabled = false;
  const router = zlink.createRouterSocket(ctx);
  const dealer = zlink.createDealerSocket(ctx);

  router.bind('inproc://flow-state-hwm-smoke');
  dealer.connect('inproc://flow-state-hwm-smoke');

  const originalSndHwm = dealer.options.sendHwm;
  dealer.options.sendHwm = 4096n;
  assert.equal(dealer.options.sendHwm, 4096n);
  dealer.options.sendHwm = originalSndHwm;

  // Idempotent RUNNING no-op transition should not disturb ordinary traffic.
  router.setReceiveFlowState(zlink.ReceiveFlowState.RUNNING);
  dealer.setReceiveFlowState(zlink.ReceiveFlowState.RUNNING);

  await dealer.send().message('flow-state-smoke').submit();
  const received = new zlink.Received();
  router.recv(received);
  assert.equal(received.parts[0].data().toString(), 'flow-state-smoke');

  await router.send(received.routingId).message('reply').submit();
  const reply = new zlink.Received();
  dealer.recv(reply);
  assert.equal(reply.parts[0].data().toString(), 'reply');

  dealer.close();
  router.close();
  ctx.close();
});

File: bindings/go/flow_state_test.go
// SPDX-License-Identifier: MPL-2.0

package zlink_test

import (
	"bytes"
	"context"
	"errors"
	"reflect"
	"strings"
	"sync"
	"testing"
	"time"

	zlink "zlink.systems/zlink"
)

// §8.1.1 "Flow-state enum 값이 C ABI 값과 같다": ZLINK_RECEIVE_FLOW_RUNNING = 0,
// ZLINK_RECEIVE_FLOW_PAUSED = 1 (core/include/zlink_enum.h).
func TestReceiveFlowStateEnumMatchesCABI(t *testing.T) {
	if zlink.ReceiveFlowRunning != zlink.ReceiveFlowState(0) {
		t.Fatalf("ReceiveFlowRunning = %d, want 0", zlink.ReceiveFlowRunning)
	}
	if zlink.ReceiveFlowPaused != zlink.ReceiveFlowState(1) {
		t.Fatalf("ReceiveFlowPaused = %d, want 1", zlink.ReceiveFlowPaused)
	}
}

// Final-review follow-up (plan §6): the three flow-state monitor event
// constants and the event-flag bits Core reuses on zlink_monitor_event_t.flags
// must be on the public root-package surface with values matching the C ABI
// (core/include/zlink_enum.h, core/include/zlink/eventing/api.h), and
// MonitorEventAll must still cover them.
func TestFlowStateMonitorEventConstantsMatchCABIAndArePublic(t *testing.T) {
	maskCases := []struct {
		name string
		got  zlink.MonitorEventMask
		want zlink.MonitorEventMask
	}{
		{"MonitorEventSendFlowPaused", zlink.MonitorEventSendFlowPaused, 1 << 16},
		{"MonitorEventSendFlowResumed", zlink.MonitorEventSendFlowResumed, 1 << 17},
		{"MonitorEventFlowStateStale", zlink.MonitorEventFlowStateStale, 1 << 18},
	}
	for _, tc := range maskCases {
		if tc.got != tc.want {
			t.Fatalf("%s = %#x, want %#x", tc.name, tc.got, tc.want)
		}
		if zlink.MonitorEventAll&tc.got == 0 {
			t.Fatalf("MonitorEventAll does not cover %s", tc.name)
		}
	}
	if zlink.MonitorEventAll != zlink.MonitorEventMask(0x7FFFF) {
		t.Fatalf("MonitorEventAll = %#x, want %#x", zlink.MonitorEventAll, 0x7FFFF)
	}

	typeCases := []struct {
		name string
		got  zlink.MonitorEventType
		want zlink.MonitorEventType
	}{
		{"MonitorEventTypeSendFlowPaused", zlink.MonitorEventTypeSendFlowPaused, 1 << 16},
		{"MonitorEventTypeSendFlowResumed", zlink.MonitorEventTypeSendFlowResumed, 1 << 17},
		{"MonitorEventTypeFlowStateStale", zlink.MonitorEventTypeFlowStateStale, 1 << 18},
	}
	for _, tc := range typeCases {
		if tc.got != tc.want {
			t.Fatalf("%s = %#x, want %#x", tc.name, tc.got, tc.want)
		}
	}

	flagCases := []struct {
		name string
		got  zlink.MonitorEventFlag
		want zlink.MonitorEventFlag
	}{
		{"MonitorEventFlagConnectionReadyEdge", zlink.MonitorEventFlagConnectionReadyEdge, 1 << 0},
		{"MonitorEventFlagSendFlowWritable", zlink.MonitorEventFlagSendFlowWritable, 1 << 1},
		{"MonitorEventFlagFlowStateStaleGeneration", zlink.MonitorEventFlagFlowStateStaleGeneration, 1 << 2},
		{"MonitorEventFlagFlowStateStaleEpoch", zlink.MonitorEventFlagFlowStateStaleEpoch, 1 << 3},
	}
	for _, tc := range flagCases {
		if tc.got != tc.want {
			t.Fatalf("%s = %#x, want %#x", tc.name, tc.got, tc.want)
		}
	}

	if zlink.MonitorTransportLaneApplication != 0 || zlink.MonitorTransportLaneCompletion != 1 {
		t.Fatalf("MonitorTransportLane values = (%d, %d), want (0, 1)",
			zlink.MonitorTransportLaneApplication, zlink.MonitorTransportLaneCompletion)
	}
}

// Final-review follow-up: the five flow-state fields must be on the public
// MonitorStatus surface (checked by field existence via the compiler; a
// runtime zero-check on a fresh PAIR socket lives in monitor_test.go).
func TestMonitorStatusExposesFlowMetricFields(t *testing.T) {
	var status zlink.MonitorStatus
	status.FlowPausedConnections = 1
	status.FlowPauseAppliedTotal = 2
	status.FlowResumeAppliedTotal = 3
	status.FlowStateStaleTotal = 4
	status.FlowPauseDurationMs = 5
	if status.FlowPausedConnections != 1 || status.FlowPauseAppliedTotal != 2 ||
		status.FlowResumeAppliedTotal != 3 || status.FlowStateStaleTotal != 4 ||
		status.FlowPauseDurationMs != 5 {
		t.Fatalf("MonitorStatus flow metric fields did not round-trip: %+v", status)
	}
}

// Spec-round follow-up: ZLINK_MONITOR_STATUS_DETAIL_FLOW_STATE
// (core/include/zlink_enum.h) must be reachable via IsFlowStateDetailPopulated()
// on the public MonitorStatus surface, agreeing bit-for-bit with DetailFlags.
func TestMonitorStatusDetailFlowStateBitMatchesCABI(t *testing.T) {
	var status zlink.MonitorStatus
	status.DetailFlags = 1 << 5
	if !status.IsFlowStateDetailPopulated() {
		t.Fatalf("IsFlowStateDetailPopulated() = false with DetailFlags = 1<<5, want true")
	}
	status.DetailFlags = 0
	if status.IsFlowStateDetailPopulated() {
		t.Fatalf("IsFlowStateDetailPopulated() = true with DetailFlags = 0, want false")
	}
}

// §8.1.1 "DEALER/ROUTER socket에서 설정이 성공하고 같은 state의 반복 호출도 성공한다".
func TestSetReceiveFlowStateSucceedsOnDealerAndRouterAndIsIdempotent(t *testing.T) {
	ctx := newContext(t)
	defer ctx.Close()

	dealer, _ := ctx.DealerSocket()
	defer dealer.Close()
	router, _ := ctx.RouterSocket()
	defer router.Close()

	for name, socket := range map[string]interface {
		SetReceiveFlowState(zlink.ReceiveFlowState) error
	}{
		"dealer": dealer,
		"router": router,
	} {
		if err := socket.SetReceiveFlowState(zlink.ReceiveFlowPaused); err != nil {
			t.Fatalf("%s SetReceiveFlowState(Paused) error = %v", name, err)
		}
		// Idempotent repeat of the same state must also succeed.
		if err := socket.SetReceiveFlowState(zlink.ReceiveFlowPaused); err != nil {
			t.Fatalf("%s SetReceiveFlowState(Paused) repeat error = %v", name, err)
		}
		if err := socket.SetReceiveFlowState(zlink.ReceiveFlowRunning); err != nil {
			t.Fatalf("%s SetReceiveFlowState(Running) error = %v", name, err)
		}
		if err := socket.SetReceiveFlowState(zlink.ReceiveFlowRunning); err != nil {
			t.Fatalf("%s SetReceiveFlowState(Running) repeat error = %v", name, err)
		}
	}
}

// §8.1.1 "PAIR, PUB/SUB 계열과 STREAM에서 not-supported 오류 mapping을 반환한다".
func TestSetReceiveFlowStateReportsNotSupportedOnUnsupportedSocketTypes(t *testing.T) {
	ctx := newContext(t)
	defer ctx.Close()

	pair, _ := ctx.PairSocket()
	defer pair.Close()
	pub, _ := ctx.PubSocket()
	defer pub.Close()
	sub, _ := ctx.SubSocket()
	defer sub.Close()
	xpub, _ := ctx.XPubSocket()
	defer xpub.Close()
	xsub, _ := ctx.XSubSocket()
	defer xsub.Close()
	stream, _ := ctx.StreamSocket()
	defer stream.Close()

	cases := map[string]interface {
		SetReceiveFlowState(zlink.ReceiveFlowState) error
	}{
		"pair":   pair,
		"pub":    pub,
		"sub":    sub,
		"xpub":   xpub,
		"xsub":   xsub,
		"stream": stream,
	}
	for name, socket := range cases {
		err := socket.SetReceiveFlowState(zlink.ReceiveFlowPaused)
		var configErr *zlink.ConfigError
		if !errors.As(err, &configErr) || configErr.Result != zlink.ConfigNotSupported {
			t.Fatalf("%s SetReceiveFlowState() error = %v, want ConfigError{Result: ConfigNotSupported}", name, err)
		}
	}
}

// §8.1.1 "Invalid handle·argument·state가 언어 오류 정책대로 mapping된다": handle case.
func TestSetReceiveFlowStateOnClosedSocketIsInvalidHandle(t *testing.T) {
	ctx := newContext(t)
	defer ctx.Close()

	dealer, _ := ctx.DealerSocket()
	if err := dealer.Close(); err != nil {
		t.Fatalf("Close() error = %v", err)
	}

	err := dealer.SetReceiveFlowState(zlink.ReceiveFlowPaused)
	var configErr *zlink.ConfigError
	if !errors.As(err, &configErr) || configErr.Result != zlink.ConfigInvalidHandle {
		t.Fatalf("SetReceiveFlowState() on closed socket error = %v, want ConfigError{Result: ConfigInvalidHandle}", err)
	}
}

// §8.1.1 "Invalid handle·argument·state가 언어 오류 정책대로 mapping된다": argument case.
func TestSetReceiveFlowStateOutOfRangeIsInvalidArgument(t *testing.T) {
	ctx := newContext(t)
	defer ctx.Close()

	dealer, _ := ctx.DealerSocket()
	defer dealer.Close()

	for _, value := range []zlink.ReceiveFlowState{2, -1, 999} {
		err := dealer.SetReceiveFlowState(value)
		var configErr *zlink.ConfigError
		if !errors.As(err, &configErr) || configErr.Result != zlink.ConfigInvalidArgument {
			t.Fatalf("SetReceiveFlowState(%d) error = %v, want ConfigError{Result: ConfigInvalidArgument}", value, err)
		}
	}
}

// §8.1.1 "Close와 경쟁해도 성공한 설정이나 close 오류 중 하나만 관찰된다".
func TestSetReceiveFlowStateRacingCloseObservesOnlyOkOrCloseRelatedError(t *testing.T) {
	ctx := newContext(t)
	defer ctx.Close()

	for i := 0; i < 20; i++ {
		dealer, _ := ctx.DealerSocket()

		var wg sync.WaitGroup
		wg.Add(2)
		var setErr error
		go func() {
			defer wg.Done()
			setErr = dealer.SetReceiveFlowState(zlink.ReceiveFlowPaused)
		}()
		go func() {
			defer wg.Done()
			_ = dealer.Close()
		}()
		wg.Wait()

		if setErr == nil {
			continue
		}
		var configErr *zlink.ConfigError
		if !errors.As(setErr, &configErr) {
			t.Fatalf("iteration %d: SetReceiveFlowState() during close error = %v, want nil or *ConfigError", i, setErr)
		}
		if configErr.Result != zlink.ConfigInvalidState && configErr.Result != zlink.ConfigInvalidHandle {
			t.Fatalf("iteration %d: SetReceiveFlowState() during close Result = %v, want InvalidState or InvalidHandle", i, configErr.Result)
		}
	}
}

// §8.1.1 "공개 표면 test에 flow frame receive·encode API가 없다": the only flow-state
// surface is the local state setter — no flow-frame receive/encode/pause-bypass API.
func TestPublicSurfaceHasNoFlowFrameAPI(t *testing.T) {
	forbidden := []string{"flowframe", "encodeflow", "receiveflowframe", "sendflowframe", "pausebypass"}

	socketTypes := []any{
		(*zlink.PairSocket)(nil),
		(*zlink.PubSocket)(nil),
		(*zlink.SubSocket)(nil),
		(*zlink.DealerSocket)(nil),
		(*zlink.RouterSocket)(nil),
		(*zlink.XPubSocket)(nil),
		(*zlink.XSubSocket)(nil),
		(*zlink.StreamSocket)(nil),
		(*zlink.CommonSocketOptions)(nil),
	}

	flowMethodsSeen := map[string]bool{}
	for _, target := range socketTypes {
		typ := reflect.TypeOf(target)
		for i := 0; i < typ.NumMethod(); i++ {
			name := typ.Method(i).Name
			lower := strings.ToLower(name)
			if strings.Contains(lower, "flow") {
				flowMethodsSeen[name] = true
			}
			for _, bad := range forbidden {
				if strings.Contains(lower, bad) {
					t.Fatalf("%v exposes forbidden flow-frame API %s", typ, name)
				}
			}
		}
	}

	want := map[string]bool{"SetReceiveFlowState": true}
	if !reflect.DeepEqual(flowMethodsSeen, want) {
		t.Fatalf("flow-related public methods = %v, want %v", flowMethodsSeen, want)
	}
}

// §8.1.1 "기존 HWM, EAGAIN 상당 결과와 send timeout 동작이 변하지 않는다": smoke that
// ordinary HWM configuration and DEALER/ROUTER traffic are unaffected by the new setter.
func TestExistingHWMBehaviorUnchangedAfterReceiveFlowStateCalls(t *testing.T) {
	ctx := newContext(t)
	defer ctx.Close()

	endpoint := inprocEndpoint("flow-state-hwm-smoke")
	router, _ := ctx.RouterSocket()
	dealer, _ := ctx.DealerSocket()
	defer router.Close()
	defer dealer.Close()

	if err := dealer.SetSendHighWaterMark(1000); err != nil {
		t.Fatalf("SetSendHighWaterMark() error = %v", err)
	}
	if got, err := dealer.SendHighWaterMark(); err != nil || got != 1000 {
		t.Fatalf("SendHighWaterMark() = (%d, %v), want (1000, nil)", got, err)
	}

	// Setting the receive-flow state (including a repeat of the default RUNNING
	// state) must not perturb byte-HWM configuration or ordinary traffic.
	if err := router.SetReceiveFlowState(zlink.ReceiveFlowRunning); err != nil {
		t.Fatalf("router SetReceiveFlowState(Running) error = %v", err)
	}
	if err := dealer.SetReceiveFlowState(zlink.ReceiveFlowRunning); err != nil {
		t.Fatalf("dealer SetReceiveFlowState(Running) error = %v", err)
	}

	if got, err := dealer.SendHighWaterMark(); err != nil || got != 1000 {
		t.Fatalf("SendHighWaterMark() after SetReceiveFlowState = (%d, %v), want (1000, nil)", got, err)
	}

	rid := zlink.NewRoutingID([]byte("flow-state-hwm-smoke-dealer"))
	_ = router.Bind(endpoint)
	_ = dealer.SetRoutingID(rid)
	_ = dealer.Connect(endpoint)
	_ = dealer.SetReceiveTimeout(5 * time.Second)

	if err := awaitRoutedSend(t, dealer.Send().Message(newMessage(t, "request")).Submit(context.Background())); err != nil {
		t.Fatalf("dealer Send() error = %v", err)
	}

	var request zlink.Received
	if _, err := router.Recv(&request, zlink.RecvFlagsNone); err != nil {
		t.Fatalf("router Recv() error = %v", err)
	}
	defer request.Close()

	if err := awaitRoutedSend(t, router.SendTo(request.RoutingID()).Message(newMessage(t, "response")).Submit(context.Background())); err != nil {
		t.Fatalf("router SendTo() error = %v", err)
	}

	var response zlink.Received
	if _, err := dealer.Recv(&response, zlink.RecvFlagsNone); err != nil {
		t.Fatalf("dealer Recv() error = %v", err)
	}
	defer response.Close()

	part, _ := response.SinglePartOrError()
	if !bytes.Equal(part.Data(), []byte("response")) {
		t.Fatalf("unexpected response = %q", string(part.Data()))
	}
}
